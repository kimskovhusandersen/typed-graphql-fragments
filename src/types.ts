export type FieldSelectionConfig<T> = T extends object
  ? {
      [P in keyof T]?: T[P] extends (infer U)[]
        ? FieldSelectionConfig<U>[] | boolean
        : FieldSelectionConfig<T[P]> | boolean;
    }
  : boolean;