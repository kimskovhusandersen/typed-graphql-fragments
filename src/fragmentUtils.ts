import { FieldSelectionConfig } from "./types";

// Overload signatures
export function generateTemplate<T>(
  config: FieldSelectionConfig<T>,
  asFieldConfig: true
): FieldSelectionConfig<T>;

export function generateTemplate<T>(
  config: FieldSelectionConfig<T>,
  asFieldConfig?: false | undefined
): string;

export function generateTemplate<T>(
  config: FieldSelectionConfig<T>,
  asFieldConfig?: boolean
): string | FieldSelectionConfig<T> {
  const buildFields = (obj: any, indentLevel: number = 0): string => {
    const indent = "  ".repeat(indentLevel);

    if (Array.isArray(obj)) {
      return obj.map((item) => buildFields(item, indentLevel)).join("\n");
    }

    return Object.entries(obj)
      .filter(([, value]) => value !== false)
      .map(([key, value]) => {
        if (Array.isArray(value)) {
          return `${indent}${key} {\n${buildFields(
            value,
            indentLevel + 1
          )}\n${indent}}`;
        } else if (typeof value === "object" && value !== null) {
          const nestedFields = buildFields(value, indentLevel + 1);
          return `${indent}${key} {\n${nestedFields}\n${indent}}`;
        } else if (value === true) {
          return `${indent}${key}`;
        }
        return "";
      })
      .filter(Boolean)
      .join("\n");
  };

  return asFieldConfig ? config : `\n${buildFields(config, 1)}\n`;
}

export function createFragment<T>(config: FieldSelectionConfig<T>) {
  return <B extends boolean | undefined = undefined>(
    asFieldConfig?: B
  ): B extends true ? FieldSelectionConfig<T> : string => {
    if (asFieldConfig) {
      return generateTemplate<T>(config, true) as any;
    } else {
      return generateTemplate<T>(config, false) as any;
    }
  };
}
