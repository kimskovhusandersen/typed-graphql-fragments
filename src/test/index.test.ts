import { createFragment } from "../fragmentUtils";
import { FieldSelectionConfig } from "../types";

interface User {
  id: number;
  name: string;
  posts: {
    title: string;
    content: string;
    comments?: {
      text: string;
      author: {
        name: string;
        email: string;
      };
    }[];
  }[];
}

describe("createFragment", () => {
  it("should handle an empty configuration", () => {
    const emptyConfig: FieldSelectionConfig<User> = {};
    const fragment = createFragment<User>(emptyConfig)();
    expect(fragment).toBe(`

`);
  });

  it("should match the snapshot for a typical User fragment", () => {
    const userConfig: FieldSelectionConfig<User> = {
      id: true,
      name: true,
      posts: [
        {
          title: true,
          content: false,
        },
      ],
    };

    const fragment = createFragment<User>(userConfig)();
    expect(fragment).toMatchSnapshot();
  });

  it("should generate a correctly formatted GraphQL fragment string", () => {
    const userConfig: FieldSelectionConfig<User> = {
      id: true,
      name: true,
      posts: [
        {
          title: true,
          content: false,
        },
      ],
    };

    const fragment = createFragment<User>(userConfig)();

    const expectedFragment = `
  id
  name
  posts {
    title
  }
`;

    expect(fragment).toBe(expectedFragment);
  });

  it("should generate a correct GraphQL fragment string for deeply nested structures", () => {
    const userConfig: FieldSelectionConfig<User> = {
      id: true,
      name: true,
      posts: [
        {
          title: true,
          content: false,
          comments: [
            {
              text: true,
              author: {
                name: true,
                email: true,
              },
            },
          ],
        },
      ],
    };

    const fragment = createFragment<User>(userConfig)();

    const expectedFragment = `
  id
  name
  posts {
    title
    comments {
      text
      author {
        name
        email
      }
    }
  }
`;

    expect(fragment).toBe(expectedFragment);
  });

  it("should generate a correct GraphQL fragment string", () => {
    const userConfig: FieldSelectionConfig<User> = {
      id: true,
      name: true,
      posts: [
        {
          title: true,
          content: false,
        },
      ],
    };

    const fragment = createFragment<User>(userConfig)();
    expect(fragment).toContain("id");
    expect(fragment).toContain("name");
    expect(fragment).toContain("posts");
    expect(fragment).toContain("title");
    expect(fragment).not.toContain("content");
  });

  it("should return a FieldSelectionConfig when asFieldConfig is true", () => {
    const userConfig: FieldSelectionConfig<User> = {
      id: true,
      name: true,
      posts: [
        {
          title: true,
          content: false,
        },
      ],
    };

    const configResult = createFragment<User>(userConfig)(true);
    expect(configResult).toEqual(userConfig);
  });
});
