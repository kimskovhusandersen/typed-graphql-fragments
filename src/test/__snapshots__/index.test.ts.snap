// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createFragment should match the snapshot for a typical User fragment 1`] = `
"
  id
  name
  posts {
    title
  }
"
`;
